### PR TITLE
Fix #532, Fix #566, add tslint in ci, add tslint/format/test/karma in precommit of git

### LINKED ISSUE #532: Run ts-lint as part of the CI
see #13316

- catch syntax errors,
- catch formatting errors (clang-format)

### LINKED ISSUE #566: Should add precommit hook to do tslint/format/karma/node test
Before we commit with git in local, we should precheck

- tslint
- format
- karma
- test-node

to make sure everything is fine.

### PR DESCRIPTION
**Pull request by JiaLiPassion**
Fix #532, Fix #566 

1. Add tslint in Travis CI
2. Modify some source codes to pass tslint.
3. Add following precommit check.
- tslint
- format
- karma singlerun with auto close ws-server
- test/node

in git pre-commit with ghooks.
So we can make sure everything is fine before we commit with git.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -22,6 +22,7 @@ before_script:
   - ./scripts/sauce/sauce_connect_block.sh
 
 script:
+  - node_modules/.bin/tslint -c tslint.json 'lib/**/*.ts' 'test/**/*.ts'
```
Comment (vicb) [inline .travis.yml]: isn't this covered by format:enforce ? 
if not what about creating a gulp task ?

File: DEVELOPER.md [hunk 2]
```diff
@@ -23,3 +23,14 @@ Run the browser tests using Karma:
...
 
 `npm run test-node`
+
+Before Commit
+------------
+
+Pre-commit hook will be executed and following check will
+automatically run before commit.
+
+- tslint
```
Comment (vicb) [inline DEVELOPER.md]: I'm -1 on creating a commit hook.
Test should run in watch mode anyway ?

File: package.json [hunk 3]
```diff
@@ -15,11 +15,19 @@
   },
   "scripts": {
     "changelog": "gulp changelog",
+    "ci": "npm run lint && npm run format && npm run test:single && npm run test-node",
```
Comment (vicb) [inline package.json]: document those in the README ?

File: .travis.yml (hunk 1 above)
Comment (JiaLiPassion) [inline .travis.yml]: Thank you for the review, you are right, there already has a task in gulp 'gulp lint', I will use that one.

File: DEVELOPER.md (hunk 2 above)
Comment (JiaLiPassion) [inline DEVELOPER.md]: Yeah, test in monitor mode can make sure all jasmine test passed.
But with pre commit, the hook can check format and lint.

Are you suggesting when code changes, automatically run format and lint check?

File: package.json (hunk 3 above)
Comment (JiaLiPassion) [inline package.json]: Yes, I will organize all those scripts and add them into README.md

**Comment by JiaLiPassion** [comment]
@vicb, I have updated the PR, and about pre-commit hook, I still keep it because I think we should check lint and format before commit. Please check whether it is a correct way or not. Or maybe we should just add lint and format in pre-commit hook and leave test alone?

File: DEVELOPER.md (hunk 2 above)
Comment (JiaLiPassion) [inline DEVELOPER.md]: ok, I see, I have removed the hook

**Comment by JiaLiPassion** [comment]
@vicb, I have removed the hook, please review, thank you.

**Pull request merged**